Add scenario about storage

yanpzhan opened the pull request:
Add scenario about storage: "attach storage is available for all workload types that are applicable".
@yapei, pls help to review for first round.

spadgett commented on frontend/integration-tests/protractor.conf.ts:
```diff
@@ -187,6 +192,7 @@ export const config: Config = {
       'tests/deploy-image.scenario.ts',
       'tests/performance.scenario.ts',
       'tests/monitoring.scenario.ts',
+      'tests/storage.scenario.ts',
```
I would move this up below secrets scenario

yanpzhan replied:
Thanks for your suggestion, will update soon.

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,55 @@
+        const item = element(by.cssContainingText('.dropdown li a', 'Add Storage'));
+        expect(item.isPresent()).toBe(true);
+        try {
+          execSync(`kubectl delete --cascade ${resource} example -n ${testName}`);
```
This should go in `afterAll`

yanpzhan replied:
Thanks for your suggestion, will update soon.

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,55 @@
+      it('check add storage item in action list', async() => {
+        await browser.wait(until.presenceOf(crudView.actionsDropdown));
+        await crudView.actionsDropdown.click();
+        const item = element(by.cssContainingText('.dropdown li a', 'Add Storage'));
```
Move the CSS selectors into `crudView`

yanpzhan replied:
Thanks for your suggestion, will update soon.

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,55 @@
+
+  resourceObjs.forEach(({kind}, resource) => {
+    describe(kind, () => {
+      it('create a new resource', async() => {
```
This seems to have a lot of common code with the curd scenario. It would be better to put this in a common helper somewhere instead of copying it here.

yanpzhan replied:
Yes, make sense,

yapei commented:
> This seems like a lot of work just to check if a menu item is there. I'm not sure how valuable this test is. It would be a lot better if we filled out the form to add a volume to the workload and checked the result.
> 
> Integration tests can be expensive to maintain, so I want to make sure we're getting good value out of the tests we add.

How about adding a `it` block checking `Attach Storage for One Resource(eg DeploymentConfig)` ?

spadgett commented:
If all we're doing is checking for the presence of the menu item, I would recommend a unit test. Here is an example:

https://github.com/openshift/console/blob/master/frontend/__tests__/components/route-pages.spec.tsx

> How about adding a it block checking Attach Storage for One Resource(eg DeploymentConfig) ? It will add value for this case

Sorry, I'm not clear on what you're proposing. Is this to fill out the Attach Storage form? I think that makes sense for an integration test, but I don't see a reason not to do it for every type we support. It should be the same code.

yanpzhan commented:
I'm considering to add " fill out the Attach Storage form" part for each type in the scenario, will have a try locally first.

spadgett commented:
By the way, here is a test the specifically tests the kebab actions if we decide to go that way:

https://github.com/openshift/console/blob/master/frontend/__tests__/components/operator-lifecycle-manager/clusterserviceversion-resource.spec.tsx#L80-L86

yapei commented:
> I'm considering to add " fill out the Attach Storage form" part for each type in the scenario, will have a try locally first.

We don't need test `fill out the Attach Storage Form` for every type

yapei commented:
> If all we're doing is checking for the presence of the menu item, I would recommend a unit test. Here is an example:
> 
> https://github.com/openshift/console/blob/master/frontend/__tests__/components/route-pages.spec.tsx
> 
> > How about adding a it block checking Attach Storage for One Resource(eg DeploymentConfig) ? It will add value for this case
> 
> Sorry, I'm not clear on what you're proposing. Is this to fill out the Attach Storage form? I think that makes sense for an integration test, but I don't see a reason not to do it for every type we support. It should be the same code.

Yes, I mean 'fill out Attach Storage Form' for one type

spadgett commented:
> We don't need test `fill out the Attach Storage Form` for every type

Any reason not to? We should be able to write the test once and loop over the types since the form is the same for all types. It should be about the same amount of code, but increases test coverage.

I'm thinking of something like this:

https://github.com/openshift/console/blob/master/frontend/integration-tests/tests/overview/overview.scenario.ts#L27-L38

yapei commented:
> Any reason not to? We should be able to write the test once and loop over the types since the form is the same for all types. It should be about the same amount of code, but increases test coverage.

Iterating over each type is much better for test coverage, but I'm a little concerned about the time consuming. It takes more time to go through all types, but they're filling the same forms.
Anyway, that's my opinion. I will follow what team decides : )

spadgett commented:
> Iterating over each type is much better for test coverage, but I'm a little concerned about the time consuming. It takes more time to go through all types, but they're filling the same forms.

It's probably only a few seconds per type. Truthfully I would be more concerned about increasing the chance of flakes since these tests gate merge.

I think we should run the entire scenario on all types or the entire scenario on a single type. I'd suggest all types for better test coverage.

yapei commented:
> I think we should run the entire scenario on all types or the entire scenario on a single type. I'd suggest all types for better test coverage.

Yes, I agree with you on this

spadgett commented:
> Yes, that's also what I'm concerning, we saw frequent and random failures when running these tests, 

OK, that's an issue even running against one type to be honest if the tests gate merge.

> and we're trying to make them more stable

Sounds good 👍

yanpzhan commented:
The scenario test seems fragile, random failure often occurs, though it passed on my local host, I will continue to try improving the stability of it.

yanpzhan commented:
/retest

yanpzhan commented:
/retest

yanpzhan commented:
/retest

yanpzhan commented:
/test all

yanpzhan commented:
/test all

yanpzhan commented:
/test all

yanpzhan commented:
/retest

yanpzhan commented:
/retest

yanpzhan commented:
/test all

yanpzhan commented:
/retest

yanpzhan commented:
/retest

yanpzhan commented:
/test all

yanpzhan commented:
@spadgett , updated the scenario, added some check steps to improve stability when add storage to  dc and deployment. Could you help to review again? Thanks!

yanpzhan commented:
@spadgett do you think if there is still some points to be updated?

yanpzhan commented:
@spadgett could you or some one review this pr again? It's hung here too long. Thanks very much!

yanpzhan commented:
/assign @rhamilto

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,62 @@
+        expect(crudView.errorMessage.isPresent()).toBe(false);
+      });
+      it(`add storage to ${resourcetype}`, async() => {
+        if (resourcetype === 'deploymentconfigs' || resourcetype === 'deployments') {
```
Why only dc and deployments? Why not all workloads?

yanpzhan replied:
During the test, I easily met conflict error only when add storage to dc/deployment, so I wait until the previous deployment finished. Seems other workloads are stable when yaml is updated.

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,62 @@
+            await browser.sleep(10000);
+          }
+          const availableCondition = $$('.row .col-xs-4').get(4);
+          const progressingCondition = $$('.row .col-xs-4').get(7);
```
These selectors are fragile. Add a `data-test-` attribute if needed.

yanpzhan replied:
Yeah, it will be good if there is data-test- attribute, seems there is no "data-test-" attribute for most elements on the detail page.

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,62 @@
+        if (resourcetype === 'deploymentconfigs' || resourcetype === 'deployments') {
+          const activeStatus = element(by.cssContainingText('.co-icon-and-text', 'Up to date'));
+          while (!activeStatus.isPresent()) {
+            await browser.sleep(10000);
```
Does this go into an infinite loop if the deployment never rolls out?

I'm not sure we need to check whether it rolls out successfully. Is it sufficient to check that the volume table has the right values? This is going to make the tests a lot slower, and I'm not sure it has a ton of value.

yanpzhan replied:
It's not suitable to use while in the test, I updated it with if, but need to make the wait time long enough.
This check is before adding storage. Since the workload is just created in previous "it" test, and sometimes adding storage would meet confliction to edit yaml. So I want to wait after the previous deployment is completed. Do you have better approach for this situation?

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,62 @@
+
+        const volumetile = element(by.cssContainingText('.co-section-heading', 'Volumes'));
+        await browser.wait(until.presenceOf(volumetile));
+        const volumebody = await $$('.co-m-table-grid__body').get(1);
```
This selector is fragile. Prefer a `data-test-` attribute if needed.

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,62 @@
+        await crudView.addNewStorageToWorkload(pvcName, pvcSize, mountPath);
+        expect(crudView.errorMessage.isPresent()).toBe(false);
+
+        const volumetile = element(by.cssContainingText('.co-section-heading', 'Volumes'));
```
Use camelCase for names

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,62 @@
+        const volumebody = await $$('.co-m-table-grid__body').get(1);
+        await browser.wait(until.presenceOf(volumebody));
+        expect(volumebody.getText()).toContain(pvcName);
+        expect(volumebody.getText()).toContain(mountPath);
```
It would be better to check that the right values show up in the right columns of the table.

yanpzhan replied:
$$('.co-m-table-grid__body .row .col-xs-5')[1]
$$('.co-m-table-grid__body .row .col-xs-7')[1]
If check the exact column with exact value, I need use above selectors which are also fragile. I wonder If there are better selector.

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -12,6 +12,11 @@ export const createYAMLLink = $('#yaml-link');
 export const saveChangesBtn = $('#save-changes');
 export const reloadBtn = $('#reload-object');
 export const cancelBtn = $('#cancel');
+export const addStorageItem = element(by.cssContainingText('.dropdown li a', 'Add Storage'));
```
We have helpers now for actions in `crud.view.ts`. We should prefer those.

yanpzhan replied:
Great. Reused clickDetailsPageAction()

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -167,3 +172,35 @@ export const emptyState = $('.cos-status-box').$('.text-center');
+  await browser.get(`${appHost}/k8s/ns/${testName}/${resourcetype}`);
+  await isLoaded();
+  const createDropdownIsPresent = await createItemButton.isPresent();
+  if (createDropdownIsPresent) {
```
I would take common creation code from `crud.scenario.ts` and add it here so we're not duplicating logic.

yanpzhan replied:
Compared code in crud.scenario.ts with this code block, Moved common code into clickYamlCreation() for reuse.

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -167,3 +172,35 @@ export const emptyState = $('.cos-status-box').$('.text-center');
+  await browser.wait(until.presenceOf(actionsButton));
+  await actionsButton.click();
+  await browser.wait(until.presenceOf(addStorageItem));
+  await addStorageItem.click();
```
There are helpers for using actions already in crud view. Use the helper.

yanpzhan replied:
Updated.

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -12,6 +12,11 @@ export const createYAMLLink = $('#yaml-link');
+export const createNewClaim = $('[value=new]');
+export const inputPVCName = $('#pvc-name');
+export const inputPVCSize = $('[name=requestSizeValue]');
+export const inputMountPath = $('#mount-path');
```
These should go in a separate storage.view.ts file

yanpzhan replied:
Good idea. Moved storage related code to storage.view.ts.

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -167,3 +172,35 @@ export const emptyState = $('.cos-status-box').$('.text-center');
+  await inputMountPath.sendKeys(mountPath);
+  await saveChangesBtn.click();
+  if ( !errorMessage.isPresent() && saveChangesBtn.isPresent() ) {
+    await saveChangesBtn.click();
```
I'm confused. Why do we need to click save twice?

yanpzhan replied:
Sometimes test would fail on the adding storage page, and I noticed that all requested items have been set, but seemed like clicking save button doesn't take effect, and there is no error info on the page. I guess if clicking save button again would avoid this issue happening. But  I'm not very sure.

yanpzhan commented:
/retest

yanpzhan commented:
/test all

yanpzhan commented:
@spadgett The ps is updated, data-test- attributions are added. Could you pls help to check again? Thanks!

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -167,3 +167,25 @@ export const emptyState = $('.cos-status-box').$('.text-center');
 export const errorMessage = $('.pf-c-alert.pf-m-inline.pf-m-danger');
 export const successMessage = $('.pf-c-alert.pf-m-inline.pf-m-success');
+
+export const clickYamlCreation = async function() {
```
I'd put more detail in the name so there's no confusion about what it does. Maybe

```suggestion
export const clickListPageCreateYAMLButton = async function() {
```

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -167,3 +167,25 @@ export const emptyState = $('.cos-status-box').$('.text-center');
+  await browser.wait(until.presenceOf(yamlView.saveButton));
+  await yamlView.saveButton.click();
+};
+
```
extra newlines

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -167,3 +167,25 @@ export const emptyState = $('.cos-status-box').$('.text-center');
+  browser.wait(until.and(untilNoLoadersPresent, until.presenceOf(element(by.cssContainingText('h1', 'Create')))));
+};
+
+export const createNamespacedResourceWithDefaultYaml = async function(resourceType:string) {
```
If this is only used in storage scenario, just inline it there.

yanpzhan replied:
This function could be reused by other scenario if they want to create a resource with default yaml.  So it may be better to put it here, WDYT?

spadgett commented on frontend/integration-tests/views/storage.view.ts:
```diff
@@ -0,0 +1,17 @@
+export const inputPVCSize = $('[name=requestSizeValue]');
+export const inputMountPath = $('#mount-path');
+
+export const addNewStorageToWorkload = async function(pvcName:string, pvcSize:string, mountPath:string) {
```
If this is only used in the storage scenario, put this there.

I'd just call this `addStorageToWorkload`

nit: space after `:`

```suggestion
export const addStorageToWorkload = async function(pvcName: string, pvcSize: string, mountPath: string) {
```

yanpzhan replied:
I name it "addNewStorageToWorkload" is to distinguish with another scenario: add existing storage to workload.  Though the latter scenario is not added yet in the repo.

> If this is only used in the storage scenario, put this there.
Are you meaning putting it in integration-tests/tests/storage.scenario.ts?

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,63 @@
+          if (!activeStatus.isPresent()) {
+            await browser.sleep(20000);
+          }
+          const availableCondition = $$('[data-test-condition="Available"] div').get(1);
```
Add a `data-test-` attribute on the exact column you want instead of just on the row.

spadgett replied:
Do we need to wait for the deployment to be available before continuing with the test? I believe we use patch, so you shouldn't see errors even if it's still rolling out.

yanpzhan replied:
It often failed in ci flow with this error, haven't you met with this error?
![Screenshot from 2019-07-03 10-40-14](https://user-images.githubusercontent.com/7313189/60567660-3e2be080-9d9d-11e9-9f04-e5da592c6f6a.png)

spadgett commented on frontend/integration-tests/tests/storage.scenario.ts:
```diff
@@ -0,0 +1,63 @@
+        const volumeRow = await $(`[data-test-volume="${pvcName}"]`);
+        await browser.wait(until.presenceOf(volumeRow));
+        expect($$(`[data-test-volume="${pvcName}"] div`).get(0).getText()).toContain(pvcName);
+        expect($$(`[data-test-volume="${pvcName}"] div`).get(1).getText()).toContain(mountPath);
```
Add `data-test-` attributes to the specific columns you need

yanpzhan commented:
/retest all

yanpzhan commented:
/test all

yanpzhan commented:
/test all

yanpzhan commented:
/test all

yanpzhan commented:
/test all

yanpzhan commented:
@spadgett Could you help to check the pr again? Thanks a lot!

yanpzhan commented:
/test all

spadgett commented on frontend/public/components/conditions.tsx:
```diff
@@ -5,11 +5,11 @@ import { Timestamp } from './utils';
       <CamelCaseWrap value={condition.type} />
     </div>
-    <div className="col-xs-4 col-sm-2 col-md-2">
+    <div className="col-xs-4 col-sm-2 col-md-2" data-test="status">
```
Prefer `data-test-id` (here and below)

spadgett commented:
> It often failed in ci flow with this error, especially when the resource is just created. Haven't you met with this error?

I don't believe this will happen anymore after this commit: https://github.com/openshift/console/commit/96183e93fc1fdf70ff2eaa0ee2510239fb89a183

Can you try rebasing and removing the extra logic to wait for the deployment to roll out?

yanpzhan commented:
@spadgett , After removed the waiting logic for dc/deployment, the test could pass without conflict error. And all new added attributes "data-test-*" now are unified with name "data-test-id". Thanks for your suggestion.

spadgett commented on frontend/integration-tests/views/crud.view.ts:
```diff
@@ -166,3 +166,23 @@ export const emptyState = $('.cos-status-box').$('.text-center');
+  browser.wait(until.and(untilNoLoadersPresent, until.presenceOf(element(by.cssContainingText('h1', 'Create')))));
+};
+
+export const createNamespacedResourceWithDefaultYaml = async function(resourceType: string) {
```
nit:

```suggestion
export const createNamespacedResourceWithDefaultYAML = async function(resourceType: string) {
```

to match case with `clickListPageCreateYAMLButton`

yanpzhan replied:
I Updated the name, but after my push, the LGTM label was removed automatically. Could you help to add the lable again? Sorry for the trouble. (Seems I could click "Commit suggestion" button directly to get your suggestion into the pr, right? not familiar that, if yes, I could try it next time.)

yanpzhan commented:
/retest

yanpzhan commented:
/retest

spadgett commented:
/retest

spadgett commented:
/lgtm cancel
build error

```
ERROR in /go/src/github.com/openshift/console/frontend/integration-tests/tests/storage.scenario.ts
(34,24): Property 'createNamespacedResourceWithDefaultYaml' does not exist on type 'typeof import("/go/src/github.com/openshift/console/frontend/integration-tests/views/crud.view")'. Did you mean 'createNamespacedResourceWithDefaultYAML'?
```

spadgett commented:
/approve
/lgtm

I pushed a fix for the build error
